examples/paper/03-dependent-binding.makam

```diff
@@ -1,90 +1,216 @@
+   λProlog is simple: before performing unification at the term-level, we perform unification at
+   the type level first, therefore further determining any uninstantiated type variables. Therefore,
+   when we check to see whether the current goal matches the premise of a rule, type unification
+   can force us to distinguish between different types. Based on these, the standard example
+   of `map` for vectors is as follows:
+*)
+
+vmap : [A B] (A -> B -> prop) -> vector N A -> vector N B -> prop.
+vmap P vnil vnil.
+vmap P (vcons X XS) (vcons Y YS) :- P X Y, vmap P XS YS.
+
+(* The notation [N] in the type of `vmap` means that the type argument `N` is ad-hoc/not-parametric.
+   Non-specified type arguments are parametric by default, so as to match standard practice in
+   languages like ML and Haskell, and to catch type errors that allowing unqualified ad-hoc
+   polymorphism would permit. For example, consider the following erroneous definition for `fold`,
+   where the arguments for `P` in the `cons` case are flipped.
+
+foldl : (B -> A -> B -> prop) -> B -> list A -> B -> prop.
+foldl P S nil S.
+foldl P S (cons HD TL) S'' <- P HD S S', foldl P S' TL S''.
+
+   If ad-hoc polymorphism is allowed for `A` and `B`, this is a well-typed definition. However, the
+   erroneous call to `P` forces the types `A` and `B` to be unified, and therefore the `fold`
+   predicate is unnecessarily restricted to only work when the two types are the same. Having to
+   specify ad-hoc polymorphism explicitly helps us avoid such mistakes. *)
+
+(* Though this support for ad-hoc polymorphism was part of the original λProlog design, we have not
+   found extensive coverage of its implications in the literature. Furthermore, it is not supported
+   well by standard implementations of λProlog (like Teyjus), which was one of the reasons that
+   prompted us to work on Makam. 
+*)
+
+(* Armed with GADTs of this form, we can now introduce dependently-typed binding forms, where the
+   number of variables that are being bound is reflected in the type. One way to do this is the
+   following: 
+*)
+
+(* dbind A N B: dependently-typed binding of N A's into a B; N will be instantiated with `natZ`
+   and `natS` as above. *)
+
+dbind : type -> type -> type -> type. 
+
+dbindbase : B -> dbind A natZ B.
+dbindnext : (A -> dbind A N B) -> dbind A (natS N) B.
 
-(* dbind A T B: dependently-typed binding of many A's into a B; T is a tuple-type representing the type of substitutions. *)
+(* Another possibility, avoiding the need for introducing type-level natural numbers, is to
+   use a more standard type as the dependent parameter: the type of tuples that would serve
+   as substitutions for the introduced variables. The type would then become:
+*)
 
 dbind : type -> type -> type -> type.
 
-%extend dbind.
+dbindbase : B -> dbind A unit B.
+dbindnext : (A -> dbind A T B) -> dbind A (A * T) B.
+
```

Comment:
I haven't understood all the issues yet, but I like the tuple version better.  It feels more right to use types to represent legit types, rather than an encoding of numbers.